### PR TITLE
minetest update

### PR DESCRIPTION
**Pull request by mridulnagpal**
@jvalleroy 
#606 
Made some changes. What do I change the configuration file for different arguments?

### CODE REVIEW

**Comment by jvalleroy** [comment]
These are the default options in /etc/minetest/minetest.conf:
```
# Maximum number of players connected simultaneously
max_users = 15

# Whether to enable players killing each other
enable_pvp = false

# Set to true to enable creative mode (unlimited inventory)
creative_mode = false

# Enable players getting damage and dying
enable_damage = true
```

If you have the augeas-tools package installed, you can check how this file is seen by Augeas (running as root user):
```
# augtool -L -A
augtool> set /augeas/load/Php/lens Php.lns
augtool> set /augeas/load/Php/incl[last()+1] /etc/minetest/minetest.conf
augtool> load
augtool> ls /files/etc/minetest/minetest.conf
.anon/ = (none)
augtool> ls /files/etc/minetest/minetest.conf/.anon/
#comment[1] = configuration file for minetestserver
...(lots of more lines)...
#comment[27] = Maximum number of players connected simultaneously
max_users = 15
#comment[28] = Whether to enable players killing each other
enable_pvp = false
#comment[29] = Set to true to enable creative mode (unlimited inventory)
creative_mode = false
#comment[30] = Enable players getting damage and dying
enable_damage = true
...(lots of more lines)...
augtool> get /files/etc/minetest/minetest.conf/.anon/max_users
/files/etc/minetest/minetest.conf/.anon/max_users = 15
augtool> get /files/etc/minetest/minetest.conf/.anon/enable_pvp
/files/etc/minetest/minetest.conf/.anon/enable_pvp = false
augtool> get /files/etc/minetest/minetest.conf/.anon/creative_mode
/files/etc/minetest/minetest.conf/.anon/creative_mode = false
augtool> get /files/etc/minetest/minetest.conf/.anon/enable_damage
/files/etc/minetest/minetest.conf/.anon/enable_damage = true
```
The values are all just strings.

This file from my pull request has examples of how to load augeas and set values in python:
https://github.com/jvalleroy/Plinth/blob/71f25826a3396a504ea4cdc9f9e470997f27877c/actions/repro

**Comment by mridulnagpal** [comment]
@jvalleroy Cool, made some changes. Please review

**Comment by mridulnagpal** [comment]
@jvalleroy Installed augeas-tools and checked it, seems to be working 
screenshot from 2016-12-19 22-38-20

**Comment by jvalleroy** [comment]
@mridulnagpal pylint found a few minor issues with code style:
```
$ pylint3 actions/minetest 
No config file found, using default configuration
************* Module minetest
C: 40, 0: Exactly one space required after comma
    configure.add_argument('--creative_mode',choices=['true', 'false'],
                                            ^ (bad-whitespace)
C: 42, 0: Exactly one space required after comma
    configure.add_argument('--enable_pvp',choices=['true', 'false'],
                                         ^ (bad-whitespace)
C: 44, 0: Exactly one space required after comma
    configure.add_argument('--enable_damage',choices=['true', 'false'],
                                            ^ (bad-whitespace)
```

If I give all 4 configure options, it can take a long time to run, because it restarts the server 4 times. Perhaps we can optimize this so it just has to restart once at the end.

Also for augeas, we can make "aug" a global variable and load it just once.

**Comment by mridulnagpal** [comment]
@jvalleroy Installed pylint3 myself will check from now on, made the changes as well

**Comment by jvalleroy** [comment]
@mridulnagpal Looks good. The next step is to add a simple configuration form to the Minetest page in Plinth. I think it will be similar to the Radicale configuration form, so you can use that as a reference.

**Comment by mridulnagpal** [comment]
@jvalleroy Added a configuration file for minetest

**Comment by jvalleroy** [comment]
@mridulnagpal Ok, next we will need to modify the MinetestServiceView:
https://github.com/freedombox/Plinth/blob/master/plinth/modules/minetest/__init__.py#L110
And have it use the new form.

**Comment by mridulnagpal** [comment]
What all modifications do I do?

**Comment by jvalleroy** [comment]
@mridulnagpal I would move it into a views.py file, and update it to look like the RadicaleServiceView.

**Comment by mridulnagpal** [comment]
@jvalleroy Created a views.py file with all the configurations and update init file as well. Please review

**Comment by jvalleroy** [comment]
@mridulnagpal Found some issues using pyflakes3:
```
jvalleroy@debian:~/fbx/plinth$ pyflakes3 plinth/modules/minetest/__init__.py 
plinth/modules/minetest/__init__.py:31: 'plinth.views.ServiceView' imported but unused
plinth/modules/minetest/__init__.py:65: undefined name 'augeas'
plinth/modules/minetest/__init__.py:65: undefined name 'augeas'
plinth/modules/minetest/__init__.py:66: undefined name 'augeas'
jvalleroy@debian:~/fbx/plinth$ pyflakes3 plinth/modules/minetest/views.py 
plinth/modules/minetest/views.py:55: undefined name 'messages'
plinth/modules/minetest/views.py:63: undefined name 'messages'
plinth/modules/minetest/views.py:71: undefined name 'messages'
plinth/modules/minetest/views.py:79: undefined name 'messages'
```

File: plinth/modules/minetest/views.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+class MinetestServiceView(ServiceView):
+    """A specialized view for configuring minetest."""
+    service_id = managed_services[0]
+    diagnostics_module_name = "minetest"
+    description = description
+    show_status_block = True
+    form_class = MinetestForm
+
+def get_initial(self):
```
Comment (jvalleroy) [inline plinth/modules/minetest/views.py]: I think these functions should be part of the MinetestServiceView class, right?

**Comment by mridulnagpal** [comment]
@jvalleroy Corrected the issues in this update. Please review

**Comment by jvalleroy** [comment]
I get an error when starting plinth:
```
jvalleroy@debian:~/fbx/plinth$ sudo ./run --debug
...
[2016-12-20 17:04:25,044] plinth.module_loader DEBUG    Importing plinth.modules.minetest
[2016-12-20 17:04:25,045] plinth.module_loader DEBUG    No URLs for minetest
Traceback (most recent call last):
  File "./run", line 21, in <module>
    plinth.__main__.main()
  File "/home/jvalleroy/fbx/plinth/plinth/__main__.py", line 338, in main
    module_loader.load_modules()
  File "/home/jvalleroy/fbx/plinth/plinth/module_loader.py", line 58, in load_modules
    _include_module_urls(module_import_path, module_name)
  File "/home/jvalleroy/fbx/plinth/plinth/module_loader.py", line 116, in _include_module_urls
    r'', django.conf.urls.include(url_module, module_name))]
  File "/usr/lib/python3/dist-packages/django/conf/urls/__init__.py", line 50, in include
    urlconf_module = import_module(urlconf_module)
  File "/usr/lib/python3.5/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 673, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "/home/jvalleroy/fbx/plinth/plinth/modules/minetest/urls.py", line 24, in <module>
    from plinth.modules.minetest import MinetestServiceView
ImportError: cannot import name 'MinetestServiceView'
```

**Comment by mridulnagpal** [comment]
@jvalleroy Do you have any idea why it is so??

**Comment by jvalleroy** [comment]
@mridulnagpal Minetest's urls.py tries to import MinetestServiceView from plinth.modules.minetest:
https://github.com/freedombox/Plinth/blob/master/plinth/modules/minetest/urls.py#L24
But since it was moved into views.py, it should be imported from plinth.modules.minetest.views.

**Comment by mridulnagpal** [comment]
@jvalleroy runs fine now

File: plinth/modules/minetest/__init__.py [hunk 2]
```diff
@@ -107,17 +122,34 @@ def disable():
...
     results = []
 
     results.append(action_utils.diagnose_port_listening(30000, 'udp4'))
 
     return results
+
+
+def get_max_players_value():
+    """Return the current Max Players value."""
+    value = aug.get(AUG_PATH + '/max_players')
```
Comment (jvalleroy) [inline plinth/modules/minetest/__init__.py]: This should be max_users, and I think we need to convert it from a string to an integer.

File: plinth/modules/minetest/__init__.py [hunk 3]
```diff
@@ -107,17 +122,34 @@ def disable():
...
+    return value
+
+
+def get_creative_mode_value():
+    """Return the current Creative mode value."""
+    value = aug.get(AUG_PATH + '/creative_mode')
```
Comment (jvalleroy) [inline plinth/modules/minetest/__init__.py]: If it's disable, the returned value will be 'false', but that's different from False. We probably want to compare the return value to the string 'true'.

**Comment by mridulnagpal** [comment]
@jvalleroy When I dont update max_players and update any other option it seems to disappear. I couldn't find the reason or the same. Please help

**Comment by jvalleroy** [comment]
@mridulnagpal The reason is that when max players is blank, it erases the value in the config:
```
# Maximum number of players connected simultaneously
max_users =
```
One thing I suggest is to keep the max players form field as an integer. Only convert it to/from a string when needed (when getting or setting the value with augeas).

**Comment by mridulnagpal** [comment]
@jvalleroy I made the changes but the problem still persists. What do I do?

File: plinth/modules/minetest/__init__.py [hunk 4]
```diff
@@ -107,17 +120,46 @@ def disable():
...
     results.append(action_utils.diagnose_port_listening(30000, 'udp4'))
 
     return results
+
+
+def get_max_players_value():
+    """Return the current Max Players value."""
+    aug = load_augeas()
+    value = aug.get(AUG_PATH + '/max_users')
+    return value
```
Comment (jvalleroy) [inline plinth/modules/minetest/__init__.py]: Why not convert to int here? It seems more natural that this function should return a number, not a string.

**Comment by mridulnagpal** [comment]
@jvalleroy I converted it to int but it still erases the value when it is not set

**Comment by jvalleroy** [comment]
I tested turning off creative mode, and this was shown in the debug log:
```
[2016-12-22 01:44:05,568] plinth.actions INFO     Executing command - ['sudo', '-n', '/usr/share/plinth/actions/minetest', 'configure', '--creative_mode', 'false']
```
So the action script is being called correctly. But in this case it clears the max_users value. So the issue must be in the action script.

I think the issue is in line 62:
```
if arguments.max_players is not False:
```
What if arguments.max_players is None, instead of False?

File: plinth/modules/minetest/views.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            value = "true"
+        else:
+            value = "false"
+
+        if get_creative_mode_value() != data['creative_mode']:
+            actions.superuser_run(
+                'minetest',
+                ['configure', '--creative_mode', value])
+            messages.success(self.request,
+                             _('Creative mode configuration updated')+char_value)
```
Comment (jvalleroy) [inline plinth/modules/minetest/views.py]: I think the char_value needs to be removed here.

**Comment by mridulnagpal** [comment]
@jvalleroy Please review.

File: actions/minetest [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    configure.add_argument('--enable_pvp', choices=['true', 'false'],
+                           help='Set player Vs player true/false')
+    configure.add_argument('--enable_damage', choices=['true', 'false'],
+                           help='Set damage true/false')
+
+    return parser.parse_args()
+
+def subcommand_configure(arguments):
+    """Configure Minetest."""
+    if arguments.max_players is not False:
```
Comment (jvalleroy) [inline actions/minetest]: We need to make a change here.

```sudo ./actions/minetest configure --creative_mode false``` should not cause the max_users setting in the config file to change. But it does, currently, it clears the value that was set.

The problem is not on the Django side: it is calling the script correctly.

**Comment by mridulnagpal** [comment]
;@jvalleroy Yes you are right I changed the script and now it works fine :+1:

**Comment by jvalleroy** [comment]
@mridulnagpal Ok, I've merged this with minor changes. Thanks!

**Pull request closed without merging**